src/main/java/com/android/volley/toolbox/DiskBasedAsyncCache.java

```diff
@@ -161,28 +174,152 @@ public void completed(Integer resultLen, Void ignore) {
                         }
                     });
         } catch (IOException e) {
             if (closeChannel(channel, "IOException")) {
                 deleteFile(file);
+                initializeIfRootDirectoryDeleted();
+            }
+            callback.onWriteComplete();
+        }
+    }
+
+    /** Clears the cache. Deletes all cached files from disk. */
+    @Override
+    public void clear(OnWriteCompleteCallback callback) {
+        File[] files = mRootDirectorySupplier.get().listFiles();
+        if (files != null) {
+            for (File file : files) {
+                deleteFile(file);
+            }
+        }
+        mEntries.clear();
+        mTotalSize = 0;
+        VolleyLog.d("Cache cleared.");
+        callback.onWriteComplete();
+    }
+
+    /**
+     * Initializes the cache. We are suppressing warnings, since we create the futures above and
+     * there is no chance this fails.
+     */
+    @Override
+    @SuppressWarnings("FutureReturnValueIgnored")
+    public void initialize(final OnWriteCompleteCallback callback) {
+        File rootDirectory = mRootDirectorySupplier.get();
+        if (!rootDirectory.exists()) {
+            createCacheDirectory(rootDirectory);
+            callback.onWriteComplete();
+            return;
+        }
+        File[] files = rootDirectory.listFiles();
+        if (files == null) {
+            callback.onWriteComplete();
+            return;
+        }
+        List<CompletableFuture<Void>> reads = new ArrayList<>();
+        for (File file : files) {
+            Path path = file.toPath();
+            AsynchronousFileChannel channel = null;
+            final int entrySize = (int) file.length();
+            final ByteBuffer buffer = ByteBuffer.allocate(entrySize);
+            final CompletableFuture<Void> fileRead = new CompletableFuture<>();
+            try {
+                channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ);
+                final AsynchronousFileChannel afc = channel;
+                afc.read(
+                        buffer,
+                        0,
+                        file,
+                        new CompletionHandler<Integer, File>() {
+                            @Override
+                            public void completed(Integer integer, File ignore) {
+                                CacheHeader entry = CacheHeader.readHeader(buffer);
+                                if (entry != null) {
+                                    closeChannel(afc, "after successful read");
+                                    entry.size = entrySize;
+                                    mTotalSize =
+                                            DiskBasedCacheUtility.putEntry(
+                                                    entry.key, entry, mTotalSize, mEntries);
+                                } else {
+                                    closeChannel(afc, "after failed read");
```

Comment:
Should you call deleteFile here?

Reply:
I didn't since it would be pruned anyway, but maybe it is better to just delete here.

Reply:
was this meant to imply you'd change it to delete? since it doesn't seem to be done. Since DiskBasedCache#initialize deletes the file in the event of any IO exception (which includes invalid headers), I think we should do the same here.